Dynamic entitlement agent

prdoyle opened the pull request:
Much rearranging to support attaching the entitlement agent to Elasticsearch.

Also includes a new `gradle run --entitlement` flag to turn this on for local development.

prdoyle committed:
Refactor: treat "maybe" JVM options uniformly

prdoyle committed:
WIP

rjernst committed:
Get entitlement running with bridge all the way through, with qualified
exports

prdoyle committed:
Cosmetic changes to SystemJvmOptions

prdoyle committed:
Disable entitlements by default

prdoyle committed:
Bridge module comments

prdoyle committed:
Fixup forbidden APIs

prdoyle committed:
spotless

prdoyle committed:
Rename EntitlementChecker

prdoyle committed:
Fixup InstrumenterTests

rjernst committed:
exclude recursive dep

prdoyle committed:
Fix some compliance stuff

prdoyle committed:
Rename asm-provider

prdoyle committed:
Stop using bridge in InstrumenterTests

prdoyle committed:
Generalize readme for asm-provider

prdoyle committed:
InstrumenterTests doesn't need EntitlementCheckerHandle

prdoyle committed:
Better javadoc

prdoyle committed:
Call parseBoolean

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into dynamic-entitlement-agent

prdoyle committed:
Add entitlement to internal module list

ldematte reviewed (commented):
Looks GREAT! 
I checked it out in Intellij and the IDE is not broken in any way, even after cleaning/reloading.
I have not run it, but I trust it works :)
Also, you definitely want to get Mark/delivery to get a final look at the gradle part.

I left some comments here and there, nothing really blocking.
One additional minor thing: I think some README need adjusting (there is still reference to `entitlement-runtime` in the agent one, and in the bridge one I think there might be imprecision/lack of detail?)

ldematte commented:
`elasticsearch-ci/part-1 ` failure seems unrelated - but please double check 👍

prdoyle commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/EntitlementInitialization.java:
```diff
@@ -0,0 +1,56 @@
+        return manager;
+    }
+
+    // Note: referenced by agent reflectively
```
Unfortunately there's no code dependency from `EntitlementInitialization` to either the bridge or agent libraries, so Javadoc won't like us making `@link` references.

prdoyle commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/EntitlementInitialization.java:
```diff
@@ -0,0 +1,56 @@
+public class EntitlementInitialization {
+    private static ElasticsearchEntitlementManager manager;
+
+    // Note: referenced by bridge reflectively
```
In the absence of `@link` or similar, I'm afraid future code evolution could mean such specifics would become stale. I'm not sure what the right approach is. 🤔

prdoyle commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementManager.java:
```diff
@@ -9,25 +9,49 @@
+public class ElasticsearchEntitlementManager implements EntitlementChecker {
     /**
-     * @return the same instance of {@link ElasticsearchEntitlementManager} returned by {@link EntitlementProvider}.
+     * Log level recommendations:
```
Oh are they? I can delete them then.

prdoyle replied:
Looking at this again... there are specific recommendations in here about what is expected at each log level, and those go beyond generic ES logger recommendations, like "_Events that happen once per instrumented method_". 

But I won't object to deleting it. I thought it made sense to capture the thinking I did when choosing these levels, but really this comment is likely to become stale over time anyway as we get more experience with troubleshooting.

prdoyle commented on libs/entitlement/asm-provider/README.md:
```diff
@@ -0,0 +1,2 @@
+This module uses the ASM library to implement various things, including bytecode instrumentation.
+It is loaded using the Embedded Provider Gradle plugin.
```
Not sure. I took it from [`ProviderLocator.load()`](https://github.com/elastic/elasticsearch/blob/6911227c058fb02084296cc548e888d6256fdbe5/libs/core/src/main/java/org/elasticsearch/core/internal/provider/ProviderLocator.java#L107).

prdoyle commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/bootstrap/EntitlementBootstrap.java:
```diff
@@ -0,0 +1,78 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class EntitlementBootstrap {
```
That's essentially a restatement of [the `bootstrap()` method's code](https://github.com/elastic/elasticsearch/blob/62a20afc0782bed219c6e3eb9c33ca28fb282fdd/libs/entitlement/src/main/java/org/elasticsearch/entitlement/bootstrap/EntitlementBootstrap.java#L28-L32).

I agree it ought to state the class's responsibilities though. Perhaps at least to say that after `bootstrap`, the entitlements are active, and any forbidden methods can't be called anymore. 🤔

ldematte replied:
I think we can point to the main doc (Markdown, in the parent folder), and that can have the image referenced (in SVG, so it's versionable too, and rendered inline by both github and Intellij)

prdoyle commented on libs/entitlement/bridge/src/main/java/org/elasticsearch/entitlement/bridge/EntitlementCheckerHandle.java:
```diff
@@ -0,0 +1,55 @@
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class EntitlementCheckerHandle {
```
I've added a little more. I worry about having each part of the code describe what the other parts do, because such comments tend to get stale over time.

prdoyle commented on libs/entitlement/asm-provider/src/main/java/module-info.java:
```diff
@@ -10,9 +10,9 @@
 import org.elasticsearch.entitlement.instrumentation.impl.InstrumentationServiceImpl;
 
-module org.elasticsearch.entitlement.agent.impl {
+module org.elasticsearch.entitlement.instrumentation {
```
Because it depends on ASM, and the main module is not allowed to do that.

ldematte replied:
That makes sense, thanks

prdoyle commented:
Good catch on the readmes @ldematte. I'll do another pass over them.

prdoyle committed:
Docs as requested by Lorenzo

prdoyle commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/EntitlementInitialization.java:
```diff
@@ -21,6 +21,13 @@
+ * instantiate and configure an {@link EntitlementChecker},
+ * make it available to the bootstrap library via {@link #checker()},
+ * and then install the {@link org.elasticsearch.entitlement.instrumentation.Instrumenter}
+ * to begin injecting our instrumentation.
```
The length of this paragraph is a hint that combining the instrumentation and entitlement checking into one library is an awkward fit. The two parts of the system are almost entirely unrelated, have distinct responsibilities, and different dependencies.

The instrumentation:
- is concerned with bytecodes
- depends on ASM
- is the same for every method (modulo argument forwarding)
- has no relation to policies / grants
- is active only during warmup
- has no steady-state throughput impact

The entitlement checking:
- is concerned with policies/grants
- doesn't use ASM
- is different for every method
- has steady-state throughput impact

There's really no overlap here, so combining these overloads one library and gives it multiple responsibilities.

ldematte replied:
I agree with what you say, but I come to different conclusions: we definitely should keep them separate "as if" they were in 2 libraries, but I think keep separate packages/classes is enough. 
It's different if we think we'll have a different implementation (non-ASM), but do you see that happen in the near future?
Anyway, I'm fine with the way it is now too.

prdoyle replied:
They ought to be separate modules because they have different dependencies. The entitlement checking module has no need for ASM; the instrumentation module has no need for Jack's policy data structures. They also have a well-defined boundary (`EntitlementChecker`). They can be tested independently.

It's an ideal place to split into two modules with cleanly defined responsibilities. I'm not sure what the hesitation is.

jdconrad commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/initialization/EntitlementInitialization.java:
```diff
@@ -0,0 +1,56 @@
+        return c.getName().replace('.', '/');
+    }
+
+    private static final InstrumentationService INSTRUMENTER_FACTORY = (new ProviderLocator<>(
```
Nit: Is there an extra set of parenthesis?

jdconrad commented on libs/entitlement/agent/src/main/java/org/elasticsearch/entitlement/agent/EntitlementAgent.java:
```diff
@@ -9,53 +9,31 @@
-import java.util.Set;
-import java.util.jar.JarFile;
 
 public class EntitlementAgent {
```
JavaDocs?

jdconrad commented on libs/entitlement/agent/src/main/java/org/elasticsearch/entitlement/agent/EntitlementAgent.java:
```diff
@@ -9,53 +9,31 @@
+    public static void agentmain(String agentArgs, Instrumentation inst) {
+        final Class<?> initClazz;
+        try {
+            initClazz = Class.forName("org.elasticsearch.entitlement.initialization.EntitlementInitialization");
```
Maybe a comment on how this is able to access `EntitlementInitialization`?

jdconrad commented on libs/entitlement/bridge/src/main/java/org/elasticsearch/entitlement/bridge/EntitlementCheckerHandle.java:
```diff
@@ -0,0 +1,58 @@
+        return Holder.instance;
+    }
+
+    private static class Holder {
```
Is this extra internal class to delay static initialization?

jdconrad commented on libs/entitlement/asm-provider/README.md:
```diff
@@ -0,0 +1,2 @@
+This module uses the ASM library to implement various things, including bytecode instrumentation.
```
Would it ever make sense to rename `asm-provider` to `instrumentation`?

prdoyle replied:
Oh the irony.

That was the original name. @rjernst suggested generalizing the name because we anticipate we'll need ASM for more than just instrumentation. I think he said you had a need for ASM?

I think, for example, we'll want to use ASM (not reflection) to scan the `EntitlementChecker` class to discover which methods to instrument. (Using reflection would force a bunch of classes to be loaded, only for them to be retransformed and reloaded again; using ASM should make startup more efficient.)

jdconrad replied:
Haha, well I'm not going to be too strongly opinionated on names so carry on :)

jdconrad commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementManager.java:
```diff
@@ -9,25 +9,49 @@
  * The trampoline module loads this object via SPI.
  */
-public class ElasticsearchEntitlementManager implements EntitlementChecks {
+public class ElasticsearchEntitlementManager implements EntitlementChecker {
```
The Elasticsearch in the name feels out of place now with all the other classes starting with Entitlement.

jdconrad replied:
My preference is to rename it, but I'd appreciate hearing @rjernst's opinion as well.

prdoyle replied:
Alright, if this works for you... I'll call it `ElasticsearchEntitlementChecker` and we can rename it again if need be once we hear Ryan's thoughts?

ldematte replied:
I'd leave it as is - renaming can be done at any phase.

prdoyle replied:
Oh I did it already. 😅

jdconrad reviewed (commented):
@prdoyle This is looking really nice! I think all the revisions have paid off in coming up with something that appears relatively simple compared to the amount of complexity (echoing @ldematte's thoughts).

Most of my comments are nits, so feel free to ignore if you'd like. I do have one thought, though, on documentation. The pieces of this are all over the place now, so JavaDocs don't really work well in this case. I think we need to be able to link back to a single source of documentation through the JavaDocs for all the important classes here to really flesh it out. Then each class's JavaDoc can just point to the single source and state what piece this is within that single source document.

prdoyle commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/bootstrap/EntitlementBootstrap.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    private static String findAgentJar() {
+        String propertyName = "es.entitlement.agentJar";
+        String propertyValue = System.getProperty(propertyName);
+        if (propertyValue != null) {
+            return propertyValue;
+        }
```
That's right. I was following a pattern @rjernst recommended from elsewhere. (I've forgotten where now...)

prdoyle replied:
TBH this gives me the creeps; it kind of looks like a security hole. But I'm sure it's fine. 😅

jdconrad replied:
I intuitively thought similarly, but logically I know if someone can set a system property were already in trouble.

ldematte replied:
> (I've forgotten where now...)

Native libs

> TBH this gives me the creeps; it kind of looks like a security hole

Same feelings :/ Don't know if this is justified, but it leaves me uneasy.
OK for now, but I'd like to review it before "going live" and see if we can do something different. Maybe we don't need this..

prdoyle committed:
Changes from Jack

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into dynamic-entitlement-agent

prdoyle committed:
Rename ElasticsearchEntitlementChecker

ldematte commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/api/ElasticsearchEntitlementChecker.java:
```diff
@@ -9,25 +9,49 @@
+public class ElasticsearchEntitlementChecker implements EntitlementChecker {
     /**
-     * @return the same instance of {@link ElasticsearchEntitlementManager} returned by {@link EntitlementProvider}.
+     * Log level recommendations:
```
Nit: I would still remove this javadoc

ldematte reviewed (approved):
Changes looks good, still one nit from the previous review but it can be done at another time.
LGTM, but I would seek approval from delivery before merging.

prdoyle commented on libs/entitlement/agent/build.gradle:
```diff
@@ -6,51 +6,18 @@
-  testImplementation project(":libs:entitlement:agent:impl")
-}
-
-tasks.named('test').configure {
```
I've deleted the tests that required this. We consider them unnecessary now, since practically nothing is actually in the agent; the logic that used to be there is in the Entitlement library now and can be tested independently.

prdoyle commented on libs/entitlement/build.gradle:
```diff
@@ -9,12 +9,25 @@
+}
+
+configurations {
+  entitlementBridge
```
I don't see any references to this. I'm not sure why it's here; perhaps left over from prior iterations of this PR.

prdoyle committed:
Remove logging javadoc

prdoyle committed:
exportInitializationToAgent should reference EntitlementInitialization, not EntitlementBootstrap.

They're currently in the same module, but if that ever changes, this code would have become wrong.

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into dynamic-entitlement-agent

mark-vieira commented on distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/SystemJvmOptions.java:
```diff
@@ -134,4 +137,32 @@ private static Stream<String> maybeWorkaroundG1Bug() {
+        }
+        String bridgeJar;
+        try (var s = Files.list(dir)) {
+            var candidates = s.limit(2).toList();
```
Why the `limit(2)` here?

prdoyle replied:
I only care if there are two or more. If there are ten thousand, I don't want to build the whole list first.

mark-vieira replied:
I think that's fine but then the following exception will be a bit odd as it will always be `found 2` regardless of how many files actually exist.

mark-vieira replied:
Also, FWIW, we're talking about code that runs exactly once, and on startup. Seems an unnecessary thing to optimize.

mark-vieira commented on distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/SystemJvmOptions.java:
```diff
@@ -12,18 +12,20 @@
-
-        return Stream.concat(
+        boolean useEntitlements = Boolean.parseBoolean(sysprops.getOrDefault("es.entitlements.enabled", "false"));
+        return Stream.of(
```
I presume at some point we'll want to stop passing `-Djava.security.manager=allow` when this is enabled. I guess it's benign to keep it for now.

prdoyle replied:
That's a good idea. I think this PR could add that.

prdoyle replied:
I've looked into it a little and that would be a nontrivial change. There are at least two other places affected:
- `initPhase1` installs a dummy `SecurityManager`
- `AllPermissionsCheck` checks that the `SecurityManager` is working

Replacing these is not difficult but it's not trivial either. Perhaps this can wait for another PR?

mark-vieira replied:
💯 No need to address in this PR. That said, we'll need to address it before we start testing on Java 24 as I presume it'll blow up if we try to pass this.

mark-vieira commented on server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java:
```diff
@@ -198,12 +199,16 @@ private static void initPhase2(Bootstrap bootstrap) throws IOException {
-            SECURITY_FILTER_BAD_DEFAULTS_SETTING.get(args.nodeSettings()),
-            args.pidFile()
-        );
+        if (Boolean.parseBoolean(System.getProperty("es.entitlements.enabled", "false"))) {
```
FWIW, this can be simplified as `Boolean.getBoolean("es.entitlements.enabled")`.

prdoyle replied:
Oh yes, in an earlier iteration, the default was `true`. But now this just looks dumb.

prdoyle replied:
Hey how about that.

```
Forbidden method invocation: java.lang.Boolean#getBoolean(java.lang.String) [use org.elasticsearch.core.Booleans#parseBoolean(java.lang.String)]
  in org.elasticsearch.bootstrap.Elasticsearch (Elasticsearch.java:202)
```

mark-vieira replied:
🤦 Ah, I actually checked to see that we use this method, and we do, but only in tests, where apparently it is not forbidden.

mark-vieira commented on libs/entitlement/build.gradle:
```diff
@@ -9,12 +9,25 @@
+  testImplementation(project(":test:framework")) {
+    exclude group: 'org.elasticsearch', module: 'entitlement'
+  }
+  entitlementBridge project(':libs:entitlement:bridge')
```
I'm not seeing where this configuration is actually used. What am I missing?

prdoyle replied:
I think you're right.

prdoyle replied:
I've looked into it a little and that would be a nontrivial change. There are at least two other places affected:
- `initPhase1` installs a dummy `SecurityManager`
- `AllPermissionsCheck` checks that the `SecurityManager` is working

Replacing these is not difficult but it's not trivial either. Perhaps this can wait for another PR?

prdoyle committed:
Some suggestions from Mark

prdoyle committed:
Turns out Boolean.getBoolean is forbidden.

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 116125`
